### ISSUE TITLE
GitHub Action: Load parent renv project when testing multiple apps via `app-dir` input

### ISSUE THREAD

**Issue by nick-youngblut** [rstudio/shinytest2#359]
I followed the example at [Using shinytest2 with continuous integration](https://rstudio.github.io/shinytest2/articles/use-ci.html), with slight modifications:

```
on:
  push:
    branches: [main, master]
  pull_request:
    branches: [main, master]

name: Test app with renv

jobs:
  test-app:
    runs-on: ${{ matrix.config.os }}

    name: ${{ matrix.config.os }} (${{ matrix.config.r }})

    strategy:
      fail-fast: false
      matrix:
        config:
          - {os: ubuntu-latest, r: "4.3.1"}

    env:
      GITHUB_PAT: ${{ secrets.GITHUB_TOKEN }}
      R_KEEP_PKG_SOURCE: yes

    steps:
      - uses: actions/checkout@v2
      - uses: r-lib/actions/setup-pandoc@v2
      - uses: r-lib/actions/setup-r@v2
        with:
          r-version: ${{ matrix.config.r }}
          use-public-rspm: true
      - uses: r-lib/actions/setup-renv@v2
      - uses: rstudio/shinytest2/actions/test-app@v1
        with:
          app-dir: |
            app
          upload-snapshots: true     
```

For the sake of GCP Cloud Run deployment, my Shiny app is located at `./app/` in the repo. 

`shinytest2` is in my renv snapshot file, and it was successfully installed in the `r-lib/actions/setup-renv@v2` step prior to `rstudio/shinytest2/actions/test-app@v1`.

My `testthat.R` file:

```
shinytest2::test_app()
```

My `setup-shinytest2.R` file:

```
# Load application support files into testing environment
shinytest2::load_app_env()
```

My `test-shinytest2.R` file:

```
library(shinytest2)

# test app initialization
test_that("App initialization is consistent", {
  app <- AppDriver$new(name = "app", height = 768, width = 1211)
  app$expect_values()
})
```

**Comment by gadenbuie** [comment]
Hi @nick-youngblut! The basic problem is that the renv isn't being loaded when the app is run from the `./app` dir. In short, this is because the script for loading the renv project lives in `./Rprofile`, which is sourced for R sessions that start in the `./` working directory (rather than the `./app` directory).

I found an answer from Kevin Ushey (renv author) [on Posit Community](https://community.rstudio.com/t/activating-renv-for-subdirectories-in-project/63463/3):

> Unfortunately, I can't think of a better method other than having a `.Rprofile` for each directory where you might want to use `renv`.

You can add an `./app/.Rprofile` file with code like the following:

```r
renv::load("../")
```

**Comment by gadenbuie** [comment]
@nick-youngblut No worries, it's not explicitly covered in the docs. In fact, looking at the section on [A repository with multiple applications](https://rstudio.github.io/shinytest2/articles/use-ci.html#a-repository-with-multiple-applications), it does seem like the intention is to support this setup without requiring user changes:

> For a repository with this structure, the configuration is largely the same as a repository with a single app, as described above. The only difference is in the `check-app.yaml` file.
> 
> `check-app.yaml` 
> The `check-app.yaml` [should] have all App dir locations supplied to `app-dir` using the multiline yaml string syntax.

@schloerke Do you think the shinytest2 test-app action should automatically load the renv in the parent directory before running the app?

https://github.com/rstudio/shinytest2/blob/176131dc1bb1bd2510100484131bb2f130dd7d99/actions/test-app/action.yml#L35-L46

**Comment by schloerke** [comment]
@gadenbuie 
This line sets up the shiny background process to load the app directory's .Rprofile / .Renviron files: https://github.com/rstudio/shinytest2/blob/176131dc1bb1bd2510100484131bb2f130dd7d99/R/app-driver-start.R#L20C1-L21

If it's within the same repo, I think it would make sense to have the individual apps's .Rprofile load the root .Rprofile. I believe this needs to be on the user (and document the work needed) or we should support it within `shinytest2::test_app(app_dir)` function.

**Comment by schloerke** [comment]
@gadenbuie I like the intent of #360, but the logic should be in `test_app()` or not done at all.  

If it's not done at all, how does local testing even work?

**Comment by schloerke** [comment]
@gadenbuie Correct

**Comment by gadenbuie** [comment]
Okay, in that case I'm not sure which way to go with this. On the one hand, if an app author is storing multiple apps in the same repo and using the same renv in each, they'd need to work out how to load the renv locally in order to run or deploy the app locally.

On the other hand, it seems much more likely that setting up multiple apps with a shared `renv.lock` file is a pattern used for testing -- i.e. the goal of the `renv.lock` is to provision a package environment for the CI runner.

Maybe a smaller change would be to set `R_LIBS_USER` to point to the renv project library in the full `check-app.yaml` workflow? What's missing in the original report is that the packages installed by renv aren't being found at the `test_app()` step.

@nick-youngblut does the repo where you noticed this problem happen to be publicly available?

**Comment by gadenbuie** [comment]
> Instead of `renv::load("../")`, could I use `source("../renv/activate.R")`?

🤦 yes if that works you should use that instead.

**Comment by nick-youngblut** [comment]
Using `source("../renv/activate.R")` did not work:

```bash
Warning: The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/
Run ## --------------------------------------------------------------------
# Bootstrapping renv 1.0.2 ---------------------------------------------------
- Downloading renv ... OK
- Installing renv  ... OK

- The project is out-of-sync -- use `renv::status()` for details.
Error: Error in loadNamespace(x) : there is no package called ‘shinytest2’
Calls: loadNamespace -> withRestarts -> withOneRestart -> doWithOneRestart
Execution halted
Error: Process completed with exit code 1.
```

It just bootstrapped an renv init.

I changed the `./app/.Rprofile` file to:

```R
source("../renv/activate.R")
renv::restore(lockfile="../renv.lock")
```

which works... except then I got the error:

```
Loading required package: shiny
✔ | F W S  OK | Context

⠏ |         0 | shinytest2                                                      
⠋ | 1       0 | shinytest2                                                      
✖ | 1       0 | shinytest2 [29.6s]
────────────────────────────────────────────────────────────────────────────────
Error (test-shinytest2.R:5:3): App initialization
Error in `app_initialize(self, private, app_dir = app_dir, ..., load_timeout = load_timeout, 
    timeout = timeout, wait = wait, expect_values_screenshot_args = expect_values_screenshot_args, 
    screenshot_args = screenshot_args, check_names = check_names, 
    name = name, variant = variant, view = view, height = height, 
    width = width, seed = seed, clean_logs = clean_logs, shiny_args = shiny_args, 
    render_args = render_args, options = options)`: Shiny app did not become stable in 15000ms.
Message: An error occurred while waiting for Shiny to be stable
Caused by error in `app_wait_for_idle()`:
! An error occurred while waiting for Shiny to be stable
```

Does my updated `./app/.Rprofile` make `r-lib/actions/setup-renv@v2` obsolete, since renv is also installed in the next step (`rstudio/shinytest2/actions/test-app@v1`)?

**Comment by gadenbuie** [comment]
@nick-youngblut It turns out that the most likely culprit for this problem is that we had updated the `actions/test-app` action but hadn't moved the sliding `v1` tag to point at the latest version of the action. I just made that change, so it's likely that your original workflow (at the start of this issue) will work now if you re-run it.

I was able to figure this out by creating a small, reproducible example: https://github.com/rstudio/shinytest2-gha-multiple-apps. If it feels unfortunate that it took this long to diagnose the problem, I would agree with you. For me this issue was a reminder of the importance of a reproducible example. I'd also note that we would have likely seen the issue much sooner if we had had access to the GHA workflow logs (in the future I'll request those sooner in the process of debugging).

We are going to move toward tagging the actions with `actions/v1` (the `v1` tag will continue to be supported) and I'll follow up with a PR to ensure our documentation is up to date.
